### PR TITLE
Allow renaming containers

### PR DESCRIPTION
**Pull request by bnjbvr**
@jankeromnes, can you take a look please? Cheers!

### CODE REVIEW

**Comment by jankeromnes** [comment]
Sure! Thanks a lot for this pull request! 😄

File: api/hosts-api.js [hunk 1]
```diff
@@ -336,7 +336,98 @@ hostAPI.get('/version', {
   }]
 });
 
-hostAPI.get('/:container/:port', {
+const containerAPI = hostAPI.api('/:container');
+
+containerAPI.patch({
+  title: 'Update a given container',
```
Comment (jankeromnes) [inline api/hosts-api.js]: Nit: Please use a shorter title like "Update a container".

File: api/hosts-api.js [hunk 2]
```diff
@@ -336,7 +336,98 @@ hostAPI.get('/version', {
...
+  description: 'Update a container\'s information.',
```
Comment (jankeromnes) [inline api/hosts-api.js]: Nit: Please use a more specific description like "Update the properties of a given Docker container.".

File: api/hosts-api.js [hunk 3]
```diff
@@ -336,7 +336,98 @@ hostAPI.get('/version', {
...
+      body: JSON.stringify({
+        name: 'Container McContaineryFace'
+      }, null, 2)
+    }
+  }]
+});
+
+containerAPI.get('/:port', {
   title: 'Get a single container port',
   description: 'Get information about a given Docker container port.',
```
Comment (jankeromnes) [inline api/hosts-api.js]: Suggestion: If you'd like to make this pre-existing description consistent with my nit above, please drive-by fix it to something like "Get the properties of a given Docker container port.".

File: api/hosts-api.js [hunk 4]
```diff
@@ -336,7 +336,98 @@ hostAPI.get('/version', {
...
+      try {
+        operations = JSON.parse(json);
+      } catch (error) {
+        response.statusCode = 400; // Bad Request
+        response.json({ error: 'Problems parsing JSON' }, null, 2);
+        return;
+      }
+
+      // FIXME: Allow real JSON Patch operations to be applied to a safe subset
+      // of the container's data (e.g. a `container.properties` object).
```
Comment (jankeromnes) [inline api/hosts-api.js]: Nit: Technically, we still call containers "machines" internally. It might make sense to reword this comment to "a safe subset of a machine's data (e.g. a `machine.properties` object)" for now.

File: templates/contributions.html [hunk 5]
```diff
@@ -12,7 +12,9 @@ <h2 class="header">My Contributions</h2>{{
           <div class="panel-heading">
             <img class="project-icon" src="{{= project.icon in xmlattr}}" alt="{{= project.name in xmlattr}} Logo">
             <h4 class="project-title">
-              {{= project.name in html}} #{{= id in id}}
+              <form action="/api/hosts/{{= machine.docker.host}}/{{= machine.docker.container}}" method="patch">
```
Comment (jankeromnes) [inline templates/contributions.html]: Nit: Even if they're not supposed to contain wacky characters, please escape the machine's `host` and `container` references, e.g. like so `{{= machine.docker.host in xmlattr}}/{{= machine.docker.container in xmlattr}}` just in case.

**Comment by nt1m** [comment]
@bnjbvr @jankeromnes I've addressed all nits except the UX one (I need to figure out how to actually test this out in dev).

**Comment by jankeromnes** [comment]
@nt1m Very cool, thanks!

Here is a hacky way to make a container show up in your contribution list. Simply kill the server, and add a `"firefox"` project and machine to your `db.json` like so:

```json
{
  "projects": {
    "firefox": {
      "id": "firefox",
      "name": "Firefox",
      "icon": "",
      "docker": {
        "host": "localhost",
        "image": "",
        "path": "",
        "ports": {
          "22": {
            "proxy": "none"
          },
          "8088": {
            "proxy": "https"
          }
        },
        "build": "",
        "update": "",
        "logs": ""
      },
      "data": {}
    }
  },
  "users": {
    "admin@localhost": {
      "machines": {
        "firefox": [
          {
            "id": 0,
            "name": "Firefox",
            "ports": {
              "ssh": 42024,
              "vnc": 42025
            },
            "docker": {
              "image": "firefox",
              "logs": "",
              "ports": {
                "22": {
                  "port": 42024,
                  "proxy": "none"
                },
                "8088": {
                  "port": 42025,
                  "proxy": "https"
                }
              },
              "container": "abcdef1234567890",
              "host": "localhost"
            },
            "status": "started",
            "data": {
              "updated": 1489079794563
            }
          }
        ]
      }
    }
  }
}
```

**Comment by nt1m** [comment]
@jankeromnes @bnjbvr I've fixed the UX, the PR should now be ready for review.

File: static/css/janitor.css [hunk 6]
```diff
@@ -280,6 +280,16 @@ h4 {
   color: #c9302c;
 }
 
+.editable-label:not(.editing) .editable-editor,
+.editable-label:not(.editing) .form-control-feedback,
+.editable-label.editing .editable-value {
```
Comment (jankeromnes) [inline static/css/janitor.css]: Nit: Maybe there is a way to capture elements less precisely, so that any new markup in this section will be also hidden/shown when useful? (E.g. `.editable-label:not(.editing) > *`).

File: templates/contributions.html [hunk 7]
```diff
@@ -11,10 +11,15 @@ <h2 class="header">My Contributions</h2>{{
         <div class="panel panel-default panel-project">
           <div class="panel-heading">
             <img class="project-icon" src="{{= project.icon in xmlattr}}" alt="{{= project.name in xmlattr}} Logo">
-            <h4 class="project-title">
-              {{= project.name in html}} #{{= id in id}}
-              <small class="project-status" data-status="{{= machine.status in xmlattr}}"></small>
-            </h4>
+            <form class="project-title editable-label ajax-form has-feedback" action="/api/hosts/{{= machine.docker.host in xmlattr}}/{{= machine.docker.container in xmlattr}}" method="patch" data-refresh-after-success="true">
+              <div class="editable-editor">
```
Comment (jankeromnes) [inline templates/contributions.html]: Nit: You can make this a form, instead of tweaking the `.project-title` thing (no need to include the `h4` in a form).

File: templates/contributions.html [hunk 8]
```diff
@@ -11,10 +11,15 @@ <h2 class="header">My Contributions</h2>{{
...
+                <input class="form-control" type="text" name="/name" value="{{= machine.properties.name in xmlattr }}" data-submit-on="blur">
+              </div>
+              <h4 class="editable-value">
+                <span>{{= machine.properties.name in xmlattr }}</span>
+                <span class="glyphicon glyphicon-pencil editable-toggle"></span>
```
Comment (jankeromnes) [inline templates/contributions.html]: Nit: This button feels pretty huge; a little too close to the text; and I think it should be hidden unless you hover over the `.project-title` element.

File: templates/contributions.html [hunk 9]
```diff
@@ -27,6 +32,9 @@ <h4 class="project-title">
             </div>
           </div>
           <div class="panel-body">
+            <label>Status</label>
+            <small class="project-status" data-status="{{= machine.status in xmlattr}}"></small>
```
Comment (jankeromnes) [inline templates/contributions.html]: Nit: Please align this better. (Also maybe it doesn't need to be a `small`, maybe a `span`?)

**Comment by jankeromnes** [comment]
(Woops, sorry for accidentally closing this pull request!)

File: api/admin-api.js [hunk 10]
```diff
@@ -103,6 +103,12 @@ oauth2providerAPI.patch({
       }
 
       // Apply the requested changes to the provider.
+      if (jsonpatch.validate(operations)) {
```
Comment (jankeromnes) [inline api/admin-api.js]: I think the correct usage of `jsonpatch.validate` is to pass it `operations` as well as the target object (i.e. `provider` in this case). And it returns an array of errors, so we should check if `.length > 0` (and maybe even log the JSON patch errors, since they should be fairly rare).

Please refer to the [jsonpatch usage](https://github.com/Starcounter-Jack/JSON-Patch#usage) documentation.

File: api/admin-api.js [hunk 11]
```diff
@@ -103,6 +103,12 @@ oauth2providerAPI.patch({
...
+        response.statusCode = 400; // Bad Request
+        response.json({ error: 'Problems parsing JSON patch' }, null, 2);
```
Comment (jankeromnes) [inline api/admin-api.js]: Nit: It's not really a problem with parsing. Maybe something more like `Invalid JSON Patch operations`?

File: api/hosts-api.js [hunk 12]
```diff
@@ -336,13 +337,120 @@ hostAPI.get('/version', {
...
+      let operations = null;
+      try {
+        operations = JSON.parse(json);
+      } catch (error) {
+        response.statusCode = 400; // Bad Request
+        response.json({ error: 'Problems parsing JSON' }, null, 2);
+        return;
+      }
+
+      if (jsonpatch.validate(operations)) {
```
Comment (jankeromnes) [inline api/hosts-api.js]: Nit: Improper use of `jsonpatch.validate` (see earlier comment).

File: api/hosts-api.js [hunk 13]
```diff
@@ -336,13 +337,120 @@ hostAPI.get('/version', {
...
+    response: {
+      body: JSON.stringify({ name: 'Much container. Such wow.' }, null, 2)
+    }
+  },{
+    request: {
+      urlParameters: {
+        hostname: 'example.com',
+        container: 'abcdef0123456789',
+      },
+      body: JSON.stringify({ name: 'Container McContaineryFace' }, null, 2)
```
Comment (jankeromnes) [inline api/hosts-api.js]: Nit: The reason this example fails is quite subtle. Maybe make it more obvious with something like `body: 'This is not a valid JSON Patch.'`.

File: api/user-api.js [hunk 14]
```diff
@@ -67,9 +67,14 @@ userAPI.patch({
         return;
       }
 
+      if (jsonpatch.validate(operations)) {
```
Comment (jankeromnes) [inline api/user-api.js]: Nit: Same as above.

File: api/user-api.js [hunk 15]
```diff
@@ -153,6 +158,12 @@ configurationsAPI.patch({
         response.json({ error: 'Problems parsing JSON' }, null, 2);
         return;
       }
+        
+      if (jsonpatch.validate(operations)) {
```
Comment (jankeromnes) [inline api/user-api.js]: Nit: Ditto.

File: lib/users.js [hunk 16]
```diff
@@ -269,6 +269,16 @@ function getOrCreateUser (email) {
...
 
+  // Temporary migration code: Previous users didn't have machine properties.
+  for (const projectId in user.machines) {
+    for (const machine of user.machines[projectId]) {
+      if (!machine.properties) {
+        machine.properties = { name: machine.name };
+        delete machine.name;
+      }
+    }
+  }
```
Comment (jankeromnes) [inline lib/users.js]: Nit: In general, we should place this temporary migration snippets after one another (to prevent an error like what happened when we tried to generate configuration files for users that didn't have names yet). However, it looks like you're lucky here and there is no dependence on earlier migration scripts.

File: package.json [hunk 17]
```diff
@@ -34,6 +34,7 @@
     "email-login": "~1.1.0",
     "fast-json-patch": "~2.0.3",
     "fleau": "~16.2.0",
+    "fsos": "^1.1.2",
```
Comment (jankeromnes) [inline package.json]: Please remove this import, as we don't use it in Janitor directly (it's already imported by email-login, for example).

**Comment by jankeromnes** [comment]
Also, please rebase and re-push this branch, which should fix the Travis tests (or at least, fail because of your actual API examples instead of an unrelated dependency error).

File: api/admin-api.js (hunk 10 above)
Comment (nt1m) [inline api/admin-api.js]: I'm not sure what you mean: https://github.com/Starcounter-Jack/JSON-Patch#jsonpatchvalidatepatch-operation-document-any-validator-function-jsonpatcherror

"If there are no errors, returns undefined. If there is an errors, returns a JsonPatchError object with the following properties"

Comment (nt1m) [inline api/admin-api.js]: Also, the target argument is optional. Not sure we want to put it ?

File: package.json (hunk 17 above)
Comment (nt1m) [inline package.json]: Whoops, not sure how that slipped in.

File: lib/users.js (hunk 16 above)
Comment (nt1m) [inline lib/users.js]: I took the order in account, and I noticed there was no dependence as well :) Not sure what would be the preferable one for you ?

**Comment by nt1m** [comment]
@nt1m: review?(@jankeromnes)

File: api/user-api.js [hunk 18]
```diff
@@ -147,14 +145,15 @@ configurationsAPI.patch({
             return;
```
Comment (jankeromnes) [inline api/user-api.js]: Nit: Doesn't the above `response.json(...)` fit in one line? If so, please drive-by fix it.

File: api/user-api.js [hunk 19]
```diff
@@ -58,18 +58,17 @@ userAPI.patch({
       json += String(chunk);
     });
     request.on('end', () => {
-      let operations = null;
       try {
-        operations = JSON.parse(json);
+        const operations = JSON.parse(json);
+        jsonpatch.applyPatch(user.profile, operations, true);
       } catch (error) {
+        log('[fail] json patch', error);
```
Comment (jankeromnes) [inline api/user-api.js]: Nit: Please make this logged message more explicit (e.g. `log('[fail] patching user.profile', error)`).

File: api/admin-api.js [hunk 20]
```diff
@@ -92,18 +92,16 @@ oauth2providerAPI.patch({
       json += String(chunk);
     });
     request.on('end', () => {
-      let operations = null;
       try {
-        operations = JSON.parse(json);
+        const operations = JSON.parse(json);
+        jsonpatch.applyPatch(provider, operations, true);
       } catch (error) {
         log('[fail] json patch', error);
```
Comment (jankeromnes) [inline api/admin-api.js]: Nit: Please drive-by fix this logged message to make it more explicit, e.g. `log('[fail] patching oauth2 provider', error)`.

File: lib/users.js [hunk 21]
```diff
@@ -273,6 +273,16 @@ function getOrCreateUser (email) {
     };
```
Comment (jankeromnes) [inline lib/users.js]: Nit: Please drive-by improve this migration code by adding a `delete user.name` line.

File: static/js/janitor.js [hunk 22]
```diff
@@ -223,6 +223,13 @@ function getFormData (form) {
   }, {});
 }
 
+// Setup editable labels.
+Array.forEach(document.querySelectorAll('.editable-label'), function (label) {
+  label.querySelector('.editable-toggle').addEventListener('click', function () {
```
Comment (jankeromnes) [inline static/js/janitor.js]: Nit: You could make this code more robust by separating out the query selector, and ensuring it actually returned an element, e.g.

```js
const toggle = label.querySelector('.editable-toggle');
if (!toggle) {
  console.error('Editable label', label, 'has no ".editable-toggle" element!');
  return;
}
```

File: templates/contributions.html [hunk 23]
```diff
@@ -11,10 +11,15 @@ <h2 class="header">My Contributions</h2>{{
         <div class="panel panel-default panel-project">
           <div class="panel-heading">
             <img class="project-icon" src="{{= project.icon in xmlattr}}" alt="{{= project.name in xmlattr}} Logo">
-            <h4 class="project-title">
-              {{= project.name in html}} #{{= id in id}}
-              <small class="project-status" data-status="{{= machine.status in xmlattr}}"></small>
-            </h4>
+            <div class="editable-label project-title">
+              <form action="/api/hosts/{{= machine.docker.host in xmlattr}}/{{= machine.docker.container in xmlattr}}" class="ajax-form editable-editor has-feedback" data-refresh-after-success="true" method="patch">
+                <input class="form-control" data-submit-on="blur" name="/name" type="text" value="{{= machine.properties.name in xmlattr }}">
```
Comment (jankeromnes) [inline templates/contributions.html]: Nit: By playing with the renaming feature, I actually lost track of which container was which. It's my fault, but it made me think, maybe would could add a `placeholder="{{= project.name in xmlattr }} #{{= id in id}}"` attribute to this input?

File: templates/contributions.html [hunk 24]
```diff
@@ -11,10 +11,15 @@ <h2 class="header">My Contributions</h2>{{
...
+              </form>
+              <div class="editable-value">
+                <h4>{{= machine.properties.name in xmlattr }}</h4>
```
Comment (jankeromnes) [inline templates/contributions.html]: Nit: The escape function should be `in html` here.

Nit: If you accept my suggestion above about the `placeholder` attribute, I think it would make sense to change this to display the default machine name when `machine.properties.name` is an empty string, e.g. like so:

```html
<h4>{{if machine.properties.name
  then {{{{= machine.properties.name in html }}}}
  else {{{{= project.name in html}} #{{= id in id}}}}}}</h4>
```

File: static/css/janitor.css [hunk 25]
```diff
@@ -332,13 +346,39 @@ h4 {
...
+}
+
+.project-title .editable-value {
+  display: flex;
+  flex-grow: 1;
+  align-items: center;
+}
+
+.panel-project:not(:hover) .editable-toggle {
+  opacity: 0;
```
Comment (jankeromnes) [inline static/css/janitor.css]: Not a nit, just a question: Why use `opacity: 0` instead of `visibility: hidden` or `display: none` on the `:hover` selector?

Nit: Please also do it for the `:focus` selector (e.g. for touch screens).

File: static/css/janitor.css [hunk 26]
```diff
@@ -280,6 +280,20 @@ h4 {
...
+  display: none;
+}
+
+.editable-editor {
+  width: 100%;
+}
+
+.editable-toggle {
+  margin-left: 10px;
+  cursor: pointer;
```
Comment (jankeromnes) [inline static/css/janitor.css]: Nit: This `.editable-toggle` icon appears black instead of the expected `#777` color. Any idea why? If you'd like, you can show it black when it is itself `:hover`ed or `:focus`ed, but not when the `.panel-project` is focused (that should only control visibility of the `.editable-toggle`, not its focus-state).

File: api/hosts-api.js [hunk 27]
```diff
@@ -336,13 +337,113 @@ hostAPI.get('/version', {
...
-const containerAPI = hostAPI.api('/:container', {
-  title: 'Containers'
+const containerAPI = hostAPI.api('/:container', 'Containers');
+
+containerAPI.patch({
+  title: 'Update a container',
+  description: 'Update the properties of a given Docker container.',
+
+  handler: (request, response) => {
+    let { user } = request;
```
Comment (nt1m) [inline api/hosts-api.js]: npm run lint didn't complain, so I guess not.

File: api/hosts-api.js [hunk 28]
```diff
@@ -336,13 +337,113 @@ hostAPI.get('/version', {
...
+    }
+
+    let json = '';
+    request.on('data', chunk => {
+      json += String(chunk);
+    });
+    request.on('end', () => {
+      try {
+        const operations = JSON.parse(json);
+        jsonpatch.applyPatch(machine.properties, operations, jsonpatch.validate);
```
Comment (nt1m) [inline api/hosts-api.js]: good catch, accidentally left this test in.

File: lib/users.js (hunk 16 above)
Comment (jankeromnes) [inline lib/users.js]: Ideally always as the last migration script in this function (safest default behavior), but you can leave this one in place if you like, since there are no obvious problems.

File: api/user-api.js (hunk 18 above)
Comment (nt1m) [inline api/user-api.js]: It does, if I put a template string instead (but not without)

File: static/css/janitor.css (hunk 25 above)
Comment (nt1m) [inline static/css/janitor.css]: no specific preference :)

Comment (nt1m) [inline static/css/janitor.css]: For the :focus selector, I suppose you meant: `.panel-project:not(:hover) .editable-toggle:not(:focus)` ? Since you can't focus a <div>
